File: src/domain/conflicts.ts
import fs from "node:fs";
import path from "node:path";
import { isProtectedRelative } from "./paths.js";

export type PlannedWrite =
  | { kind: "create"; relative: string }
  | { kind: "overwrite"; relative: string }
  | { kind: "skip"; relative: string; reason: string }
  | { kind: "gitignore"; action: "create" | "append" | "skip" }
  | { kind: "db"; action: "create" | "migrate"; path: string };

export function classifyFilePlan(
  targetDir: string,
  relative: string,
  force: boolean,
): Extract<PlannedWrite, { kind: "create" | "overwrite" | "skip" }> {
  const abs = path.join(targetDir, relative);
  const exists = fs.existsSync(abs);
  if (!exists) {
    return { kind: "create", relative };
  }
  if (force) {
    return { kind: "overwrite", relative };
  }
  if (isProtectedRelative(relative)) {
    return {
      kind: "skip",
      relative,
      reason: "protected path exists (use --force to overwrite with backup)",
    };
  }
  return {
    kind: "skip",
    relative,
    reason: "already exists",
  };
}

export function hasBlockingConflicts(
  plans: PlannedWrite[],
  force: boolean,
): string[] {
  if (force) return [];
  return plans
    .filter(
      (p): p is Extract<PlannedWrite, { kind: "skip" }> =>
        p.kind === "skip" && p.reason.includes("protected"),
    )
    .map((p) => p.relative);
}

File: src/domain/conflicts.rs
use std::path::Path;

use super::paths::is_protected_relative;

#[derive(Clone, Debug)]
pub enum PlannedWrite {
    Create { relative: String },
    Overwrite { relative: String },
    Skip { relative: String, reason: String },
    Gitignore { action: String },
    Db { action: String, path: String },
}

pub fn classify_file_plan(target_dir: &Path, relative: &str, force: bool) -> PlannedWrite {
    let abs = target_dir.join(relative);
    if !abs.exists() {
        return PlannedWrite::Create {
            relative: relative.to_string(),
        };
    }
    if force {
        return PlannedWrite::Overwrite {
            relative: relative.to_string(),
        };
    }
    if is_protected_relative(relative) {
        PlannedWrite::Skip {
            relative: relative.to_string(),
            reason: "protected path exists (use --force to overwrite with backup)".into(),
        }
    } else {
        PlannedWrite::Skip {
            relative: relative.to_string(),
            reason: "already exists".into(),
        }
    }
}

pub fn blocking_conflicts(plans: &[PlannedWrite], force: bool) -> Vec<String> {
    if force {
        return Vec::new();
    }
    plans
        .iter()
        .filter_map(|p| match p {
            PlannedWrite::Skip { relative, reason } if reason.contains("protected") => {
                Some(relative.clone())
            }
            _ => None,
        })
        .collect()
}
